packages/ui/src/components/popover/popover.tsx

```diff
@@ -0,0 +1,198 @@
+ *          <i className="icon" />
+ *      </button>
+ */
+
+export class Popover extends Component<PopoverProps, PopoverState> {
+  private _closeTimeout: any = null;
+  private _controlElement = null;
+  static defaultProps = {
+    ...defaultProps,
+  };
+  state = {
+    open: false,
+  };
+
+  private _clearTimeout = () => {
+    clearTimeout(this._closeTimeout);
+    this._closeTimeout = null;
+  };
+
+  private _handleClickOutside = (e: any) => {
+    if (
+      !!this._controlElement &&
+      !this._controlElement.contains(e.target) &&
+      this.state.open
+    ) {
+      this._closePopover();
+    }
+  };
+
+  private _openPopover = () => {
+    const {onOpen} = this.props;
+    this.setState({open: true}, () => {
+      if (onOpen) {
+        onOpen();
+      }
+    });
+  };
+
+  private _closePopover = () => {
+    const {onClose} = this.props;
+    this._clearTimeout();
+    this.setState({open: false}, () => {
+      if (onClose) {
+        onClose();
+      }
+    });
+  };
+
+  private _handleEscButtonPressed = (e: KeyboardEvent) => {
+    // check if ESC or Enter pressed
+    if (this.state.open && (e.keyCode === 27 || e.keyCode === 13)) {
+      this._closePopover();
+    }
+  };
+  private _handleClick = () => {
+    if (this.state.open) {
+      this._closePopover();
+    } else {
+      this._openPopover();
+    }
+  };
+  private _handleMouseEnter = () => {
+    if (!this.state.open) {
+      this._openPopover();
+    }
+  };
+  private _handleMouseLeave = () => {
+    this._closeTimeout = setTimeout(this._closePopover, CLOSE_ON_HOVER_DELAY);
+  };
+  private _handleHoverOnPopover = () => {
+    if (this.state.open && this._closeTimeout) {
+      this._clearTimeout();
+    } else {
+      this._closePopover();
+    }
+  };
+  private _getHoverEvents = () => {
+    if (this.props.triggerMode === PopoverTriggerMode.Hover) {
+      return {
+        targetEvents: {
+          onMouseEnter: this._handleMouseEnter,
+          onMouseLeave: this._handleMouseLeave,
+        },
+        popoverEvents: {
+          onMouseEnter: this._handleHoverOnPopover,
+          onMouseLeave: this._handleHoverOnPopover,
+        },
+      };
+    }
+    return {targetEvents: {onClick: this._handleClick}, popoverEvents: {}};
+  };
+  render(props: PopoverProps): JSX.Element | null {
+    if (!props.content || !props.children) {
+      return null;
+    }
+    const {targetEvents, popoverEvents} = this._getHoverEvents();
+    const verticalAlignment =
+      props.verticalPosition === PopoverVerticalPositions.Top
+        ? PopoverVerticalPositions.Top
+        : PopoverVerticalPositions.Bottom;
+    const horizontalAlignment =
+      props.horizontalPosition === PopoverHorizontalPositions.Left
+        ? PopoverHorizontalPositions.Left
+        : PopoverHorizontalPositions.Right;
+    return (
+      <div className={styles.popoverContainer}>
+        <div
+          className="popover-anchor-container"
+          ref={node => {
```

Comment:
does toggling works on the icon?

Reply:
It's not clear. What do U mean under "icon toggling"?